push-payload/index.js

```diff
@@ -1,34 +1,57 @@
-var endpoint;
-var key;
-var authSecret;
+var mySubscription;
+
+// This function is needed because Chrome doesn't accept a base64 encoded string
+// as value for applicationServerKey in pushManager.subscribe yet
+// https://bugs.chromium.org/p/chromium/issues/detail?id=802280
+function urlBase64ToUint8Array(base64String) {
+  const padding = '='.repeat((4 - base64String.length % 4) % 4);
+  const base64 = (base64String + padding)
+    .replace(/\-/g, '+')
+    .replace(/_/g, '/');
+ 
+  const rawData = window.atob(base64);
+  const outputArray = new Uint8Array(rawData.length);
+ 
+  for (let i = 0; i < rawData.length; ++i) {
+    outputArray[i] = rawData.charCodeAt(i);
+  }
+  return outputArray;
+}
 
 // Register a Service Worker.
 navigator.serviceWorker.register('service-worker.js')
 .then(function(registration) {
   // Use the PushManager to get the user's subscription to the push service.
   return registration.pushManager.getSubscription()
-  .then(function(subscription) {
-    // If a subscription was found, return it.
+  .then(async function(subscription) {
+    // Get the server's public key
+    var response = await fetch('./vapidPublicKey')
+    var vapidPublicKey = await response.text();
+    // Convert the URL safe base64 string to a Uint8Array
+    var convertedVapidKey = urlBase64ToUint8Array(vapidPublicKey);
+
+    // See if we already have a subscription with the server
     if (subscription) {
-      return subscription;
+      var currentKey = new Uint8Array(subscription.options.applicationServerKey);
+      if (currentKey.toString() == convertedVapidKey.toString()) {
+        // The server is still using the same encryption keys, so we can keep our existing subscription
+        return subscription;
+      } else {
+        // The server has changed its encryption keys, so we need to setup a new subscription
+        await subscription.unsubscribe();
+      }
     }
 
-    // Otherwise, subscribe the user (userVisibleOnly allows to specify that we don't plan to
+    // Subscribe the user (userVisibleOnly allows to specify that we don't plan to
     // send notifications that don't have a visible effect for the user).
-    return registration.pushManager.subscribe({ userVisibleOnly: true });
+    return registration.pushManager.subscribe({
+      userVisibleOnly: true,
+      applicationServerKey: convertedVapidKey
+    });
   });
 }).then(function(subscription) {
-  // Retrieve the user's public key.
-  var rawKey = subscription.getKey ? subscription.getKey('p256dh') : '';
-  key = rawKey ?
-        btoa(String.fromCharCode.apply(null, new Uint8Array(rawKey))) :
-        '';
-  var rawAuthSecret = subscription.getKey ? subscription.getKey('auth') : '';
-  authSecret = rawAuthSecret ?
-               btoa(String.fromCharCode.apply(null, new Uint8Array(rawAuthSecret))) :
-               '';
-
-  endpoint = subscription.endpoint;
+  // Keep the subscription for easy access
+  mySubscription = subscription;
```

Comment:
We shouldn't need this, we can use `subscription` directly.

Reply:
It's reused at line 63 in the doIt.onclick function. If we don't serve it here we need to add the following there (fine with me either way, but this adds some more code):

```javascript
navigator.serviceWorker.ready.then(function(registration) {
    registration.pushManager.getSubscription().then(async function(subscription) {
        ...
    }
}
````

Reply:
We can bring doIt.onclick in the scope here, it needs to wait for the subscription anyway.